### PR TITLE
Fix: Gemma 3 & 4 Base Model and Flax Linen Decoding  

### PR DESCRIPTION
**Pull request by RexBearIU**
# Description
  
This PR introduces architectural configurations for Gemma3 and Gemma4, and implements critical decoding layer KV-cache propagation fixes for JAX/NNX
  (`nnx_decoders.py`) and Flax Linen (`decoders.py`) pipelines.
  
Before this change, Flax Linen scanned blocks (`_apply_gemma3_scanned_blocks` and `_apply_gemma4_scanned_blocks`) lacked mechanisms to propagate    
  intermediate KV-caches across scanned layers under `nn.scan`, causing end-to-end decodes to fail.
  
Key Changes:
    - **Flax Linen Decoder (`src/maxtext/layers/decoders.py`)**: Added `kv_caches` and `attention_metadata` propagation across Gemma3/4 scanned blocks  
  via a stack/unstack mapping over `nn.scan`.
    - **JAX/NNX Decoder (`src/maxtext/layers/nnx_decoders.py`)**: Unified dynamic KV-cache carry updates within scanned block execution.
    - **Model Scaffolding (`gemma3.py`, `gemma4.py`)**: Added `kv_cache` routing through scannable blocks and layer inputs.
  
# Tests
  
- Validated via end-to-end decoding with Gemma3-4B on an 8-device TPU mesh.
- Run unit tests:
```bash
python3 -m pytest tests/post_training/unit/lora_utils_test.py
```
- Decode test:
- [command and output](https://paste.googleplex.com/4521857989607424)

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/utils/maxtext_utils.py [hunk 1]
```diff
@@ -2067,3 +2067,37 @@ def get_mesh_from_config(
...
+  if kv_caches is None or returned_kv_cache is None:
+    return
+
+  if stacked:
+    # Unstack the returned tuple of stacked PyTrees back into a list of tuples
+    unstacked = [jax.tree_util.tree_map(lambda x, i=i: x[i], returned_kv_cache) for i in range(scan_length)]
+    for i in range(scan_length):
+      start_idx = i * block_len
+      for offset, updated_item in enumerate(unstacked[i]):
+        kv_caches[start_idx + offset] = updated_item
```
Comment (NuojCheng) [inline src/maxtext/utils/maxtext_utils.py:2103]: is kv_caches a python list, not jax array? jax array is not mutable. If so, use if isinstance(kv_caches) List to make sure kv_caches is a list

**Review by NuojCheng — APPROVED** [review]

File: src/maxtext/utils/maxtext_utils.py (hunk 1 above)
Comment (RexBearIU) [reply src/maxtext/utils/maxtext_utils.py:2103]: Yes, good catch. `kv_caches` is indeed a Python list containing JAX PyTrees. Since JAX arrays/PyTrees are immutable but we mutate the outer list structure in `update_kv_caches_after_scan`, I've added explicit `isinstance(kv_caches, list)` checks and validation tests to ensure it behaves correctly. I just pushed the fixes.

**Review by khatwanimohit — APPROVED** [review]
